src/meta/src/barrier/mod.rs

```diff
@@ -231,149 +235,139 @@ impl CheckpointControl {
-        prev_epoch: u64,
-        result: Vec<BarrierCompleteResponse>,
-    ) -> Vec<EpochNode> {
-        // change state to complete, and wait for nodes with the smaller epoch to commit
-        let wait_commit_timer = self.metrics.barrier_wait_commit_latency.start_timer();
-        if let Some(node) = self
-            .command_ctx_queue
-            .iter_mut()
-            .find(|x| x.command_ctx.prev_epoch.value().0 == prev_epoch)
-        {
-            assert!(matches!(node.state, InFlight));
-            node.wait_commit_timer = Some(wait_commit_timer);
-            node.state = Completed(result);
-        };
-        // Find all continuous nodes with 'Complete' starting from first node
-        let index = self
-            .command_ctx_queue
-            .iter()
-            .position(|x| !matches!(x.state, Completed(_)))
-            .unwrap_or(self.command_ctx_queue.len());
-        let complete_nodes = self.command_ctx_queue.drain(..index).collect_vec();
-        complete_nodes
-    }
-
-    /// Remove all nodes from queue and return them.
-    fn barrier_failed(&mut self) -> Vec<EpochNode> {
-        self.command_ctx_queue.drain(..).collect_vec()
-    }
-
     /// Pause inject barrier until True.
     fn can_inject_barrier(&self, in_flight_barrier_nums: usize) -> bool {
-        let in_flight_not_full = self
-            .command_ctx_queue
-            .iter()
-            .filter(|x| matches!(x.state, InFlight))
-            .count()
-            < in_flight_barrier_nums;
+        let in_flight_not_full = self.inflight_command_ctx_queue.len() < in_flight_barrier_nums;
 
         // Whether some command requires pausing concurrent barrier. If so, it must be the last one.
         let should_pause = self
-            .command_ctx_queue
+            .inflight_command_ctx_queue
             .back()
-            .map(|x| x.command_ctx.command.should_pause_inject_barrier())
+            .map(|command| command.command_ctx.command.should_pause_inject_barrier())
             .unwrap_or(false);
         debug_assert_eq!(
-            self.command_ctx_queue
+            self.inflight_command_ctx_queue
                 .iter()
-                .any(|x| x.command_ctx.command.should_pause_inject_barrier()),
+                .any(|command| command.command_ctx.command.should_pause_inject_barrier()),
             should_pause
         );
 
         in_flight_not_full && !should_pause
     }
 
-    /// Check whether the target epoch is managed by `CheckpointControl`.
-    pub fn contains_epoch(&self, epoch: u64) -> bool {
-        self.command_ctx_queue
-            .iter()
-            .any(|x| x.command_ctx.prev_epoch.value().0 == epoch)
-    }
-
     /// We need to make sure there are no changes when doing recovery
-    pub fn clear_changes(&mut self) {
-        self.finished_jobs.clear();
+    pub async fn clear_and_fail_all_nodes(&mut self, err: &MetaError) {
+        // join spawned completing command to finish no matter it succeeds or not.
```

Comment:
IIUC, after this PR, there will be no pending join handle if we reach here because `failure_recovery` will only be triggered when `next_complete_barrier` is returned with an error. At that time, the pending join handle must be finish and `self.completing_command` must be None.

That being said, should we assert here?

Reply:
I have made some change on the code. 

In current code, we will call `failure_recovery` when any barrier fails to be collected, which can be at anytime and may happen when there is some pending spawned `complete_barrier` task unfinished yet.